stream-chat-android-client/src/main/java/io/getstream/chat/android/client/scope/user/UserJob.kt

```diff
@@ -0,0 +1,127 @@
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:OptIn(InternalCoroutinesApi::class)
+@file:Suppress("DEPRECATION_ERROR")
+
+package io.getstream.chat.android.client.scope.user
+
+import io.getstream.chat.android.client.models.UserId
+import io.getstream.logging.StreamLog
+import kotlinx.coroutines.ChildHandle
+import kotlinx.coroutines.ChildJob
+import kotlinx.coroutines.CompletableJob
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A user aware implementation of [Job].
+ */
+internal interface UserJob : CompletableJob {
+    /**
+     * Cancels all children of the [UserJob] in this context connected to the specified [userId].
+     */
+    fun cancelChildren(userId: UserId? = null)
+}
+
+/**
+ * Creates a user aware job object in an active state.
+ */
+internal fun UserJob(parent: Job? = null, getUserId: () -> UserId?): UserJob = UserJobImpl(
+    SupervisorJob(parent), getUserId
+)
+
+/**
+ * A user aware implementation of [Job]. It is optionally a child to a parent job.
+ */
+private class UserJobImpl(
+    private val delegate: CompletableJob,
+    private val getUserId: () -> UserId?
+) : CompletableJob by delegate, UserJob {
+
+    private val logger = StreamLog.getLogger("Chat:UserJob")
+
+    /**
+     * Wraps child job with [UserChildJob] and attaches it so that [UserJob] job becomes its parent.
+     *
+     * WARNING! This function is marked as [InternalCoroutinesApi].
+     */
+    override fun attachChild(child: ChildJob): ChildHandle {
+        val userId = getUserId()
+        return delegate.attachChild(UserChildJob(userId, child))
+    }
+
+    /**
+     * Cancels all children of the [UserJob] in this context connected to the specified [userId].
+     */
+    override fun cancelChildren(userId: UserId?) {
+        logger.d { "[cancelChildren] userId: '$userId'" }
+        for (child in children) {
+            if (child is UserChildJob && child.userId != userId && userId != null) {
+                logger.v { "[cancelChildren] skip child: $child)" }
+                continue
+            }
+            logger.v { "[cancelChildren] cancel child: $child)" }
+            child.cancel()
+        }
+    }
+
+    /**
+     * Returns the element with the given [key] from this context or `null`.
+     *
+     * Intentionally overridden to use [super] and not a [delegate] one.
+     */
+    override fun <R> fold(initial: R, operation: (R, CoroutineContext.Element) -> R): R {
```

Comment:
Is this `override` needed? I believe this is just calling the super class version of `fold`. Same question for `get` and `minusKey`

Reply:
Hey @leandroBorgesFerreira ,

Yes, it's needed, because `UserJobImpl` delegates everything to `delegate`:
```kotlin
private class UserJobImpl(
    private val delegate: CompletableJob,
    private val getUserId: () -> UserId?
) : CompletableJob by delegate, UserJob {
```
However, we need to use `super` implementation and  not a `delegate` one